src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/DefaultRazorLanguageServerCustomMessageTarget.cs

```diff
@@ -1304,4 +1313,79 @@ public override Task<ImplementationResult> ImplementationAsync(DelegatedPosition
     }
 
     private record struct DelegationRequestDetails(string LanguageServerName, Uri ProjectedUri, ITextBuffer TextBuffer);
+
+    private class TelemetryReportingLSPRequestInvoker : LSPRequestInvoker
+    {
+        private LSPRequestInvoker _requestInvoker;
+        private ITelemetryReporter _telemetryReporter;
+        public TelemetryReportingLSPRequestInvoker(LSPRequestInvoker requestInvoker, ITelemetryReporter telemetryReporter)
+        {
+            _requestInvoker = requestInvoker;
+            _telemetryReporter = telemetryReporter;
+        }
+
+        public override Task<IEnumerable<ReinvokeResponse<TOut>>> ReinvokeRequestOnMultipleServersAsync<TIn, TOut>(string method, string contentType, TIn parameters, CancellationToken cancellationToken)
+        {
+            return _requestInvoker.ReinvokeRequestOnMultipleServersAsync<TIn, TOut>(method, contentType, parameters, cancellationToken);
+        }
+
+        public override Task<IEnumerable<ReinvokeResponse<TOut>>> ReinvokeRequestOnMultipleServersAsync<TIn, TOut>(string method, string contentType, Func<JToken, bool> capabilitiesFilter, TIn parameters, CancellationToken cancellationToken)
+        {
+            return _requestInvoker.ReinvokeRequestOnMultipleServersAsync<TIn, TOut>(method, contentType, capabilitiesFilter, parameters, cancellationToken);
+        }
+
+        public override IAsyncEnumerable<ReinvocationResponse<TOut>> ReinvokeRequestOnMultipleServersAsync<TIn, TOut>(ITextBuffer textBuffer, string method, TIn parameters, CancellationToken cancellationToken)
+        {
+            return _requestInvoker.ReinvokeRequestOnMultipleServersAsync<TIn, TOut>(textBuffer, method, parameters, cancellationToken);
+        }
+
+        public override IAsyncEnumerable<ReinvocationResponse<TOut>> ReinvokeRequestOnMultipleServersAsync<TIn, TOut>(ITextBuffer textBuffer, string method, Func<JToken, bool> capabilitiesFilter, TIn parameters, CancellationToken cancellationToken)
+        {
+            return _requestInvoker.ReinvokeRequestOnMultipleServersAsync<TIn, TOut>(textBuffer, method, capabilitiesFilter, parameters, cancellationToken);
+        }
+
+        public override Task<ReinvokeResponse<TOut>> ReinvokeRequestOnServerAsync<TIn, TOut>(string method, string languageServerName, TIn parameters, CancellationToken cancellationToken)
+        {
+            using (_telemetryReporter?.BeginBlock(nameof(ReinvokeRequestOnServerAsync), Severity.Normal, ImmutableDictionary<string, object?>.Empty
+                    .Add("eventscope.method", method)
```

Comment:
❓ would a pooled builder make sense here? @DustinCampbell

Reply:
I wouldn't think so. Since it's just an ImmutableDictionary with one key/value pair, a builder would be overkill. Just adding that single element off of the static `Empty` field is about the most efficient way to get what you want, allocation-wise. A builder essentially just does the same thing, though with a bit more overhead.

Reply:
Oh, sorry. I only saw the line that was being commented on and assumed that only a single item was being added.

Reply:
I suppose a pooled dictionary builder would be preferrable here in that case! :smile: